### ISSUE TITLE
we need to automate it to convert spaces to tabs

### ISSUE THREAD

**Issue by erfangnu** [SalamLang/Salam#164]
we need to automate it to convert spaces to tabs

https://github.com/SalamLang/Salam/commit/bbf56b1e8a6cf6c4e8d49f87dc95ce245bf5612e

**Comment by jbampton** [comment]
refs #46

**Comment by BaseMax** [comment]
Please check it mate @jbampton

**Comment by BaseMax** [comment]
After merging @BDadmehr0 changes, now everything is spaces. We need to change all to Tabs.

**Comment by BDadmehr0** [comment]
I ran clang-format once and the codes are cleaned, but this happens automatically before commit

**Comment by BaseMax** [comment]
Please add all requirements and useful commands to REAMDE or somewhere.
And if it doesn't run automatically in pre-commit, we can add cang in pre-commit settings.

**Comment by BaseMax** [comment]
It also seems that we need the same linter for python files. https://github.com/SalamLang/Salam/issues/214

**Comment by BDadmehr0** [comment]
I'm working on it.

**Comment by BDadmehr0** [comment]
> Please add all requirements and useful commands to REAMDE or somewhere. And if it doesn't run automatically in pre-commit, we can add CLANG in pre-commit settings.

I created a README.md about this. I just have a question: where should I add or upload this README.md?

**Comment by BaseMax** [comment]
CONTRIBUTING
Maybe https://github.com/SalamLang/Salam/blob/main/CONTRIBUTING.md

**Comment by BDadmehr0** [comment]
> CONTRIBUTING Maybe https://github.com/SalamLang/Salam/blob/main/CONTRIBUTING.md

I have read the CONTRIBUTING guide, but I couldn't find any information about where to upload the documentation for common commands and the prerequisites for clang-format.

**Comment by BaseMax** [comment]
Yes, we don't have it, but we can edit the CONTRIBUTING file to create and add useful information about the development and format of clang and other tools in that file.

Feel free to change anything you want.

**Comment by BDadmehr0** [comment]
https://github.com/SalamLang/Salam/pull/227

**Comment by BaseMax** [comment]
Thanks

**Comment by BaseMax** [comment]
This issue is related to an automated process that converts all space indents to tabs. Have we already implemented this feature in the pre-commit stage?

**Comment by BDadmehr0** [comment]
> This issue is related to an automated process that converts all space indents to tabs. Have we already implemented this feature in the pre-commit stage?

Check the [pre-commit configuration files](https://github.com/SalamLang/Salam/blob/main/.pre-commit-config.yaml) or git hooks(.git/hooks/) to see if the process for converting spaces to tabs is implemented.

**Comment by BaseMax** [comment]
@jbampton clear answer please

**Comment by jbampton** [comment]
Python black is using spaces for indents

**Comment by jbampton** [comment]
https://github.com/psf/black/issues/47

@BaseMax

**Comment by BaseMax** [comment]
it's better to use tabs for all. agree?

**Comment by BaseMax** [comment]
any updates?

**Comment by BaseMax** [comment]
Any updates?

**Comment by BDadmehr0** [comment]
Why can't this issue be closed? Is there a problem? @jbampton  @BaseMax

**Comment by BaseMax** [comment]
cc @jbampton

**Comment by BaseMax** [comment]
- Add a CLI flag (e.g. --tabs, or --indent=tab|4|2) threaded from driver_fmt into fmt_tokens.
- In the loop at line 59, emit either one '\t' per indent level or indent * width spaces based on that flag.
- Add a tests/fmt/* case asserting tab output + idempotency.

**Comment by BaseMax** [comment]
Yes, we are fully supporting it well:

```bash
# format source in place (auto-detects nothing - pass --lang=fa for Persian files)
salamc fmt app.salam                            # reformat one file
salamc fmt                                      # reformat every .salam under the cwd, recursively
salamc fmt src/ examples/                       # reformat given files and/or directories
salamc fmt --check                              # report files that need formatting (exit 1 if any)
salamc fmt page.salam --lang=fa                 # Persian source
```

**Comment by BaseMax** [comment]
https://github.com/SalamLang/Salam/pull/1183

New commands:

```bash
# format source in place (auto-detects nothing - pass --lang=fa for Persian files)
salamc fmt app.salam                            # reformat one file
salamc fmt                                      # reformat every .salam under the cwd, recursively
salamc fmt src/ examples/                       # reformat given files and/or directories
salamc fmt --check                              # report files that need formatting (exit 1 if any)
salamc fmt app.salam --tabs                     # indent with tabs (convert spaces to tabs)
salamc fmt src/ --indent=2                      # indent with 2 spaces per level
salamc fmt page.salam --lang=fa                 # Persian source
```
